More tests

povilasb opened the pull request:
Adds an integration test that checks how quinn deals with many connections via a single endpoint.
This test is slow and thus by default is run when "slow_tests" feature is specified. 
Note, that travis run tests with "--all-features".

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,197 @@
+#![cfg(feature = "slow_tests")]
```
Would [the `#[ignore]` attribute](https://doc.rust-lang.org/book/ch11-02-running-tests.html#ignoring-some-tests-unless-specifically-requested) be a better way to accomplish this? Along with adding `--ignored` to the CI script.

djc replied:
I think features are more appropriate for this, why would you prefer ignore?

Ralith replied:
The documentation example is for this exact use-case, and it ensures that the test is still typechecked and built as part of the normal workflow.

djc replied:
What documentation example do you mean here? I don't think we have any other tests that really exercise this specific use case (including large payloads).

Plus, I would value making the added dependencies optional conditional on the feature flag.

povilasb replied:
Indeed docs do say that:
```
 Rather than listing as arguments all tests you do want to run, you can instead annotate the time-consuming tests using the ignore attribute to exclude them

When you’re at a point where it makes sense to check the results of the ignored tests and you have time to wait for the results, you can run cargo test -- --ignored instead.
```

The only problem I see with `ignored` is that I will have to decorate helper functions as unused.

Ralith replied:
> What documentation example do you mean here?

The example(s) in the documentation for the ignore attribute: https://doc.rust-lang.org/book/ch11-02-running-tests.html#ignoring-some-tests-unless-specifically-requested

> The only problem I see with ignored is that I will have to decorate helper functions as unused.

`#![allow(unused)]` at the top of the module, perhaps?

Ralith replied:
Wait, why would you have to decorate anything as unused? The function is still compiled and referenced, just not invoked at runtime without a flag; that's still a static use.

povilasb replied:
Good point, updated the PR ;)

povilasb commented:
I think so. I rebased the PR and updated the test so that when message  is transferred via the connection, server closes this connection, then client side is expected to notice connection close and in such case client connection driver is expected to complete as well.

povilasb commented:
That of course does not test all scenarios of correct background task completion, e.g.

* connection close, should terminate local end connection driver
* connection close, should terminate remote connect driver
* connection close, should terminate incoming streams task, etc.

I'm planning to make a separate PR for those test cases - I have the barebones code.

povilasb commented:
Btw, this test occasionally just freezes, which indicates another potential bug. I can reproduce this more reliably with http://github.com/povilasb/quinn-tests and I believe this is not a test code issue. Hopefully will have more info later.

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,203 @@
+    let wait_for_all_connections = conn_done_rx
+        .take(ctx.borrow().expected_messages as u64)
+        .for_each(|_| Ok(()));
+    unwrap!(runtime.block_on(wait_for_all_connections));
```
Could [`Runtime::run`](https://docs.rs/tokio/0.1.18/tokio/runtime/current_thread/struct.Runtime.html#method.run) be used here instead of explicit message-passing?

povilasb replied:
It's a neat idea, thanks. Although I will still have to notify endpoint driver that I'm done with the test, cause it is the last spawned task blocking `Runtime::run()` :)

Ralith replied:
Could you accomplish that without explicit messaging by dropping the `Incoming` stream (along with all other `Endpoint` handles) after the expected number of connections have been handled?

Either way, it's nice to invoke `run` here to ensure that there aren't any stale futures left floating around, since the shutdown logic is nontrivial.

povilasb replied:
hm, `Incoming` and `EndpointDriver` are spawned on the event loop, so I don't have handles to them anymore. Besides signaling them with messaging, I don't know how I could drop those tasks :/

Ralith replied:
You know how many connections you're expecting, so you can use `Stream::take` to finish the `Incoming` stream after that number; once the resulting future is finished, the runtime will automatically drop it. `EndpointDriver` will automatically finish (and be dropped) once `Incoming` and all `Endpoint` values have been dropped and all `Connection`s are complete.

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,203 @@
+    let mut peer_cfg = peer_cfg_builder.build();
+    let transport_config = unwrap!(Arc::get_mut(&mut peer_cfg.transport));
+    transport_config.idle_timeout = 0;
+    transport_config.keep_alive_interval = 10_000;
```
Why disable the idle timeout? Leaving it at something reasonable might be useful for converting the test timeout into a failure. Additionally, keep-alive isn't useful if the idle timeout is disabled.

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,200 @@
+            });
+        runtime.spawn(task);
+    }
+    drop(endpoint); // we don't need it anymore, this will make EndpointDriver finish.
```
```suggestion
    drop(endpoint); // we don't need it anymore, this will make EndpointDriver finish after all connections are cleaned up```

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,200 @@
+    expected_messages: usize,
+
+    /// Hold connections so that streams wouldn't be closed prematurely
+    connections: Vec<quinn::Connection>,
```
This shouldn't be necessary; if dropping a `Connection` interferes with its still-active streams or `IncomingStreams`, that's a bug.

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,200 @@
+    unwrap!(peer_cfg_builder.add_certificate_authority(their_cert));
+    let mut peer_cfg = peer_cfg_builder.build();
+    let transport_config = unwrap!(Arc::get_mut(&mut peer_cfg.transport));
+    transport_config.idle_timeout = 30_0000;
```
```suggestion
    transport_config.idle_timeout = 30_000;
```
Unless it actually needs 300 seconds.

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,200 @@
+use tokio::runtime::current_thread::{self, Runtime};
+use unwrap::unwrap;
+
+struct TestContext {
```
Do we actually need this struct at all any more?

Ralith commented on quinn/tests/many_connections.rs:
```diff
@@ -0,0 +1,200 @@
+            if ctx.received_messages == ctx.expected_messages {
+                // let done_tx = unwrap!(ctx.test_done_tx.take());
+                // unwrap!(done_tx.send(()));
+            }
```
Dead code

povilasb commented:
Oops, that was a rushed push. Thanks for the thorough review - I addressed all your comments now.

djc commented:
This was merged as #304.
